### PR TITLE
filesystem: convert free functions to object methods

### PR DESCRIPTION
**Pull request by sesmith177**
*Description*:

Move the free functions in Envoy::Filesystem to the Filesystem::Instance object. Pass the API object containing the Instance object through everywhere filesystem access is needed. Update `canonicalPath` to return `SysCallStringResult` instead of throwing an exception as requested in feedback on https://github.com/envoyproxy/envoy/pull/5470#discussion_r248820627 

This is step 1/3 in breaking up #5470 to add Windows filesystem support as outlined here: https://github.com/envoyproxy/envoy/pull/5470#discussion_r248847967

This PR also overlaps a bit with #5660, so we should probably get one of them through first

cc @jmarantz @mattklein123 @lizan

*Risk Level*:
Low, though this change touches many files.

*Testing*:
`bazel build //source/... && bazel test //test/...`

*Docs Changes*:
N/A
*Release Notes*:
N/A

### CODE REVIEW

File: include/envoy/filesystem/watcher.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Envoy {
+namespace Filesystem {
+
+/**
+ * Abstraction for a file watcher.
+ */
+class Watcher {
+public:
+  typedef std::function<void(uint32_t events)> OnChangedCb;
```
Comment (jmarantz) [inline include/envoy/filesystem/watcher.h]: Prefer 'using' to 'typedef'.

File: test/common/json/config_schemas_test.cc [hunk 2]
```diff
@@ -26,10 +27,16 @@ std::vector<std::string> generateTestInputs() {
   return file_list;
 }
 
-class ConfigSchemasTest : public ::testing::TestWithParam<std::string> {};
+class ConfigSchemasTest : public ::testing::TestWithParam<std::string> {
+public:
```
Comment (jmarantz) [inline test/common/json/config_schemas_test.cc]: general pattern is to use 'protected' here in a test class.

File: test/common/json/json_loader_test.cc [hunk 3]
```diff
@@ -10,8 +11,16 @@
 namespace Envoy {
 namespace Json {
 
-TEST(JsonLoaderTest, Basic) {
-  EXPECT_THROW(Factory::loadFromFile("bad_file"), Exception);
+class JsonLoaderTest : public testing::Test {
+public:
```
Comment (jmarantz) [inline test/common/json/json_loader_test.cc]: protected, here & below...I won't point out additional instances of that.

**Comment by jmarantz** [comment]
and you'll need to merge master of course.

**Comment by mattklein123** [comment]
@sesmith177 thanks for slogging through this. This is going to be great for us in the future. Can you merge master and we can get this merged?

**Comment by sesmith177** [comment]
@mattklein123 @jmarantz merged

**Pull request merged**